src/main/java/org/folio/linked/data/service/resource/meta/MetadataServiceImpl.java

```diff
@@ -0,0 +1,35 @@
+package org.folio.linked.data.service.resource.meta;
+
+import static java.util.Objects.nonNull;
+import static org.folio.ld.dictionary.ResourceTypeDictionary.INSTANCE;
+import static org.folio.linked.data.model.entity.ResourceSource.LINKED_DATA;
+
+import java.util.UUID;
+import org.folio.linked.data.model.entity.InstanceMetadata;
+import org.folio.linked.data.model.entity.Resource;
+import org.springframework.stereotype.Service;
+
+@Service
+public class MetadataServiceImpl implements MetadataService {
+
+  @Override
+  public void ensureMetadata(Resource resource) {
+    ensureMetadata(resource, null);
+  }
+
+  @Override
+  public void ensureMetadata(Resource resource, InstanceMetadata oldResourceMeta) {
+    if (resource.isOfType(INSTANCE)) {
+      var metadata = new InstanceMetadata(resource)
+        .setResource(resource)
+        .setSource(LINKED_DATA);
+      if (nonNull(oldResourceMeta)) {
+        metadata.setInventoryId(oldResourceMeta.getInventoryId());
+        metadata.setSrsId(oldResourceMeta.getSrsId());
+      } else {
+        metadata.setInventoryId(UUID.randomUUID().toString());
+      }
+      resource.setInstanceMetadata(metadata);
+    }
+  }
```

Comment:
minor
suggest to always remove "full method cover `if` statmets

instead of it use return after condition

```java
 if (isNotInstance(resource)) {
    return
}
//and now all the buisness code at the main level, not part of if statement

var metadata = new InstanceMetadata(resource)
...
  
```

Reply:
sure, done